apiserver/facades/client/application/application.go

```diff
@@ -731,6 +742,133 @@ func (api *APIBase) SetCharmProfile(args params.ApplicationSetCharmProfile) erro
 	return application.SetCharmProfile(args.CharmURL)
 }
 
+// WatchLXDProfileUpgradeNotifications returns a watcher that fires on LXD
+// profile events.
+func (api *APIBase) WatchLXDProfileUpgradeNotifications(entity params.Entity) (params.NotifyWatchResult, error) {
+	var nothing params.NotifyWatchResult
+	if err := api.checkCanRead(); err != nil {
+		return nothing, errors.Trace(err)
+	}
+
+	tag, err := names.ParseApplicationTag(entity.Tag)
+	if err != nil {
+		return nothing, errors.Trace(err)
+	}
+
+	app, err := api.backend.Application(tag.Id())
+	if err != nil {
+		return nothing, errors.Trace(err)
+	}
+	w, err := app.WatchLXDProfileUpgradeNotifications()
+	if err != nil {
+		return nothing, errors.Trace(err)
+	}
+
+	_, ok := <-w.Changes()
+	if !ok {
+		return params.NotifyWatchResult{
+			Error: common.ServerError(watcher.EnsureErr(w)),
+		}, nil
+	}
+	return params.NotifyWatchResult{
+		NotifyWatcherId: api.resources.Register(w),
+	}, nil
+}
+
+// GetLXDProfileUpgradeMessages returns the lxd profile messages associated with
+// a application and set of machines
+func (api *APIBase) GetLXDProfileUpgradeMessages(arg params.LXDProfileUpgradeMessages) (results params.LXDProfileUpgradeMessagesResults, err error) {
+	defer func() {
+		if err != nil {
+			if resErr := api.resources.Stop(arg.WatcherId); resErr != nil {
+				logger.Errorf("unable stop resource %+v", resErr)
+			}
+		}
+	}()
+
+	if err = api.checkCanRead(); err != nil {
+		return
+	}
+
+	var tag names.Tag
+	tag, err = names.ParseApplicationTag(arg.ApplicationTag)
+	if err != nil {
+		return
+	}
+	var app Application
+	app, err = api.backend.Application(tag.Id())
+	if err != nil {
+		return
+	}
+	var units []Unit
+	units, err = app.AllUnits()
+	if err != nil {
+		return
+	}
+
+	finished := true
+	results = params.LXDProfileUpgradeMessagesResults{
+		Results: make([]params.LXDProfileUpgradeMessagesResult, len(units)),
+	}
+	for k, unit := range units {
+		// always set the unit name on the result.
+		results.Results[k].UnitName = unit.Tag().String()
+
+		machineId, err := unit.AssignedMachineId()
```

Comment:
this is good.  a subordinate unit will return the principals machine id with AssignedMachineId()